### DISCUSSION TITLE
DynamoDB - contract for a binary attribute value

### DISCUSSION THREAD

**Discussion by andrykonchin** [aws/aws-sdk-ruby#3146] [Q&A] [answered]
I've just found out that a value of a binary item attribute could be specified as a `StringIO` instance, like it's mentioned in the example:

```
dynamodb.put_item(
  table_name: 'aws-sdk',
  item: {
    id: 'uuid',
    age: 35,
    tags: Set.new(%w(simple attributes)),
    data: StringIO.new('data'),
    scores: [5, 4.5, 4.9, nil],
    name: {
      first: 'John',
      last: 'Doe',
    }
  }
)
```

https://github.com/aws/aws-sdk-ruby/blob/version-3/gems/aws-sdk-dynamodb/lib/aws-sdk-dynamodb/plugins/simple_attributes.rb#L20-L33

Similarly binary value loaded from DynamoDB is represented with a `StringIO` instance.

So the question is the following. Is `StringIO` should be considered as a public API (until some breaking changes are introduced)? And consequently only `StringIO` instances should be provided in requests and cannot be used any other `IO`-like classes (e.g. `File`). Or public API is just `IO` class' "interface". Or maybe even more strictly - only the `#read` method?

**Comment by mullermp** [comment, marked as answer]
Thanks for opening a discussion. When marshalling, an IO or StringIO is acceptable. When I marshalling, it comes out as a StringIO.

https://github.com/aws/aws-sdk-ruby/blob/version-3/gems%2Faws-sdk-dynamodb%2Flib%2Faws-sdk-dynamodb%2Fattribute_value.rb#L67

What are your concerns with not using these classes? I believe File is also an IO.

**Reply by andrykonchin** [reply]
Thank you for so quick response.

I have no problems or concerns with IO and StringIO classes.

But as far as it's not documented (or at least I haven't found any mentioning of IO/StringIO classes) I was wondering what can be changed and what cannot in the current major SDK version. For instance can we rely on returning StringIO and use its non-IO methods like `StringIO#string` or not (looks like we can). Can a not-IO-subclass but rather a class that implements required public interface, e.g. the `#read` method, be used in a request (looks like cannot)?

Just out of curiosity (I don't have any problem with it) but I am wondering if the `StringIO#read` method is used in the SDK to build a request. I haven't profiled it (yet, my bad) but looking at its [source code](https://github.com/ruby/ruby/blob/v3_3_6/ext/stringio/stringio.c#L1590) I would say it makes a copy of an underlined String, what means an additional memory allocation that can be avoided with e.g. `StringIO#string`.

**Reply by mullermp** [reply]
Thanks for your response. The original link you provided does show binary data having a response of StringIO, but I do think the documentation can be improved. If you open an issue on our repo or would like to improve that, we can go from there. As far as relying on StringIO, I think you can safely assume that StringIO will be the response and we cannot change that in the major version, so you can use the `string` method if you prefer there.

As far as request building goes, we copy the body (which is a StringIO) to the Net::HTTP body stream here: https://github.com/aws/aws-sdk-ruby/blob/e36fb440966a3f736d2456cc378c05956664668f/gems/aws-sdk-core/lib/seahorse/client/net_http/handler.rb#L157-L160

**Reply by andrykonchin** [reply]
Thank you!

Sorry for such delay. Created an issue https://github.com/aws/aws-sdk-ruby/issues/3153.
